### PR TITLE
Implement scBasset motif injection procedure

### LINKED ISSUE #1952: Add the motif_score method of scBasset in scvi-tools
Thank you for your effort on implement ScBasset.

Can a similar approach to `scbasset.utils.motif_score` also be integrated into scvi-tools in the future?

The implementation in scbasset is very simple, the [-8] layer of the model is used as the output of the model and then simple arithmetic is performed. 

I want to try to extract the corresponding layer from scvi's model (torch) like keras for prediction, but I failed, I can't find where the model parameters are stored. So is it possible to add the `motif_score` function in the next version?

method link: https://github.com/calico/scBasset/blob/c15bec3a73fa1e04822db723338d234ca9d384ce/scbasset/utils.py#L453

```Python
def motif_score(tf, model, motif_fasta_folder, bc=False):
    """score motifs for any given TF.
    Args:
        tf:             TF of interest. By default we only provide TFs to score in
                        https://storage.googleapis.com/scbasset_tutorial_data/Homo_sapiens_motif_fasta.tar.gz.
                        To score on additional motifs, follow make_fasta.R in the tarball 
                        to create dinucleotide shuffled sequences with and without motifs of
                        interest.
        model:          a trained scBasset model.
        motif_fasta_folder: folder for dinucleotide shuffled sequences with and without any motif.
                        We provided motifs from CIS-BP/Homo_sapiens.meme downloaded from the
                        MEME Suite (https://meme-suite.org/meme/) in 
                        https://storage.googleapis.com/scbasset_tutorial_data/Homo_sapiens_motif_fasta.tar.gz.
    Returns:
        array:          a vector for motif activity per cell. (cell order is the
                        same order as the model.)
    """
    fasta_motif = "%s/shuffled_peaks_motifs/%s.fasta" % (motif_fasta_folder, tf)
    fasta_bg = "%s/shuffled_peaks.fasta" % motif_fasta_folder

    pred_motif = pred_on_fasta(fasta_motif, model, bc=bc)
    pred_bg = pred_on_fasta(fasta_bg, model, bc=bc)
    tf_score = pred_motif.mean(axis=0) - pred_bg.mean(axis=0)
    tf_score = (tf_score - tf_score.mean()) / tf_score.std()
    return tf_score
```

```Python
def pred_on_fasta(fa, model, bc=False, scale_method=None):
    """Run a trained model on a fasta file.
    Args:
        fa:             fasta file to run on. Need to have a fixed size of 1344. Default
                        sequence size of trained model.
        model:          a trained scBasset model.
    Returns:
        array:          a peak*cell imputed accessibility matrix. Sequencing depth corrected for.
    """
    records = list(SeqIO.parse(fa, "fasta"))
    seqs = [str(i.seq) for i in records]
    seqs_1hot = np.array([dna_1hot(i) for i in seqs])
    pred = imputation_Y_normalize(seqs_1hot, model, bc_model=bc, scale_method=scale_method)
    return pred
```

```Python
# perform imputation. Depth normalized.
def imputation_Y_normalize(X, model, bc_model=False, scale_method=None):
    """Perform imputation. Normalize for depth.
    Args:
        X:              feature matrix from h5.
        model:          a trained scBasset model.
    Returns:
        array:          a peak*cell imputed accessibility matrix. Sequencing depth corrected
                        for. scale_method=None, don't do any scaling of output. The raw
                        normalized output would have both positive and negative values.
                        scale_method="all_positive" scales the output by subtracting minimum value.
                        scale_method="sigmoid" scales the output by sigmoid transform.
    """
    if bc_model:
        new_model = tf.keras.Model(
            inputs=model.layers[0].input,
            outputs=model.layers[-8].output,
        )
        Y_pred = new_model.predict(X)
        w = model.layers[-6].get_weights()[0]
        accessibility_norm = np.dot(Y_pred.squeeze(), w)
    
    else:
        new_model = tf.keras.Model(
            inputs=model.layers[0].input,
            outputs=model.layers[-4].output,
        )
        Y_pred = new_model.predict(X)
        w = model.layers[-3].get_weights()[0]
        accessibility_norm = np.dot(Y_pred.squeeze(), w)

    # scaling
    if scale_method == "positive":
        accessibility_norm = accessibility_norm - np.min(accessibility_norm)
    if scale_method == "sigmoid":
        accessibility_norm = np.divide(
            1,
            1 + np.exp(-accessibility_norm),
        )
    
    return accessibility_norm

```

### PR DESCRIPTION
**Pull request by jacobkimmel**
Fixes #1952 

- Adds the scBasset motif injection procedure to the `SCBASSET` model.
- **Note**: During implementation, I realized that the `SCBASSET` implementation here differs from the original. The original uses a fixed length input of `1344` bp, whereas `scvi-tools` used `1334`. This seems to stem from a slight different in max pooling behavior. There's a simple fix to restore the original behavior if desired.
- I have not changed to match the original implementation, but rather crop the motifs prior to injection to match the change in fixed input length.

**Added dependencies**

- Motif injection libraries are downloaded from the original manuscript as local FASTA files. `biopython` is used to read the FASTA files. An optional dependency was added in `pyproject` under `regseq`.

**Tests**

- Tests were added for all additional methods introduced by the PR. 
- An integration test was performed mimicking the `SCBASSET` tutorial. TF motifs presented in Figure 5 of the original manuscript were scored in the 10X Multiome PBMC dataset. Results appear to match expectations, though without ground truth, confirming an exact reproduction of the original implementation remains a **TODO**

**Manuscript result**

image

**This implementation, 10X PBMC data**

image

### CODE REVIEW

**Comment by adamgayoso** [comment]
Thanks @jacobkimmel ! 

The 1334 must've been an early typo/bug and some compensation with the pooling to make it seem correct.

We should change the default here and then keep your changes as default to match the original exactly.

https://github.com/scverse/scvi-tools/blob/2cf00ecef4a04dfa2d35fb0fd3cb3aa0eb101330/scvi/data/_preprocessing.py#L376-L388

File: scvi/external/scbasset/_model.py
Comment (adamgayoso) [inline scvi/external/scbasset/_model.py:369]: should we be chunking this into minibatches?

File: scvi/external/scbasset/_model.py
Comment (adamgayoso) [inline scvi/external/scbasset/_model.py:214]: we have an internal download function we can use here:

https://github.com/jacobkimmel/scvi/blob/a3ac84016ad2f30e481edd944c1cf65e2f00a4cd/scvi/data/_download.py#L13-L14

File: scvi/external/scbasset/_model.py
Comment (adamgayoso) [inline scvi/external/scbasset/_model.py:242]: no need for types here again

File: scvi/external/scbasset/_model.py
Comment (adamgayoso) [reply scvi/external/scbasset/_model.py:315]: also I think the returns here should just be `tf_score`?

File: scvi/external/scbasset/_model.py
Comment (adamgayoso) [inline scvi/external/scbasset/_model.py:275]: might be best to import biopython things in this function

we have a dependencies decorator here

https://github.com/jacobkimmel/scvi/blob/a3ac84016ad2f30e481edd944c1cf65e2f00a4cd/scvi/_decorators.py#L18-L19

This can be used like `@dependencies("biopython")`

File: pyproject.toml
Comment (adamgayoso) [inline pyproject.toml:87]:
```suggestion
regseq = [
    "biopython>=1.81",
    "genomepy",
]
```

File: scvi/external/scbasset/_model.py
Comment (adamgayoso) [inline scvi/external/scbasset/_model.py:288]:
```suggestion
        motif_dir: Optional[str] = None,
```

File: scvi/external/scbasset/_model.py
Comment (adamgayoso) [inline scvi/external/scbasset/_model.py:302]:
```suggestion
        lib_size_norm : bool, optional
```

**Review by adamgayoso — COMMENTED** [review]
Thanks for making this PR! Most comments are small, but I think the biggest thing we should do is change everything so `seq_len=1344` is the default

File: scvi/external/scbasset/_module.py
Comment (jacobkimmel) [reply scvi/external/scbasset/_module.py:69]: 👍

File: tests/external/test_scbasset.py
Comment (jacobkimmel) [reply tests/external/test_scbasset.py:43]: oh cool, fixed

File: scvi/external/scbasset/_model.py
Comment (jacobkimmel) [reply scvi/external/scbasset/_model.py:315]: 1. will remove types
2. Yes, I think co-pilot went a lil crazy and I didn't catch it

File: scvi/external/scbasset/_model.py
Comment (jacobkimmel) [reply scvi/external/scbasset/_model.py:345]: +1, used this implementation. nice one-liner.

File: scvi/external/scbasset/_model.py
Comment (jacobkimmel) [reply scvi/external/scbasset/_model.py:369]: yep, fixed

File: scvi/external/scbasset/_model.py
Comment (jacobkimmel) [reply scvi/external/scbasset/_model.py:242]: it pains me to remove types, but I'll do it for you

File: scvi/external/scbasset/_model.py
Comment (jacobkimmel) [reply scvi/external/scbasset/_model.py:214]: switched

File: scvi/external/scbasset/_model.py
Comment (jacobkimmel) [reply scvi/external/scbasset/_model.py:275]: also very cool, added

File: scvi/external/scbasset/_model.py
Comment (adamgayoso) [reply scvi/external/scbasset/_model.py]: they will get rendered on readthedocs! Though won't be there in an IDE (but can see the signature on an IDE easily)

**Comment by jacobkimmel** [comment]
Comments should be addressed in the new commits. Thanks for the quick review and glad to have scBasset as part of scvi. big thanks to the whole team for a great tool!

File: scvi/external/scbasset/_model.py
Comment (jacobkimmel) [reply scvi/external/scbasset/_model.py]: I jest, should be cleaned up :)

File: tests/external/test_scbasset.py
Comment (adamgayoso) [inline tests/external/test_scbasset.py:50]:
```suggestion
@pytest.mark.internet
```

File: tests/external/test_scbasset.py
Comment (adamgayoso) [inline tests/external/test_scbasset.py:41]:
```suggestion
@pytest.mark.internet
```

File: tests/external/test_scbasset.py
Comment (adamgayoso) [inline tests/external/test_scbasset.py:1]:
```suggestion
import pytest
from pathlib import Path
```

**Review by adamgayoso — APPROVED** [review]
LGTM, tests are failing due to a new anndata issue with solo.

I will merge this anyway. @jacobkimmel do you have an interest in updating the tutorials? Either way is fine!

**Comment by jacobkimmel** [comment]
Happy to update the tutorials. Is best practice to do so after merge so that:

```
import scvi_colab
install()
```

works without pointing to this branch, which will probably get deleted soon?

**Comment by adamgayoso** [comment]
Yes probably best to wait until this gets merged, unless you want to do it locally, which I anticipate is much faster than Colab.

File: scvi/external/scbasset/_model.py
Comment (adamgayoso) [inline scvi/external/scbasset/_model.py:356]: Actually, last thing, this isn't applicable anymore right?

Comment (jacobkimmel) [reply scvi/external/scbasset/_model.py:356]: Oh yeah, missed this one. Replaced with a warning to users if they decide to load motifs from another source that get trimmed, throws `ValueError` if motif sizes are smaller than the input size of the model.

**Pull request merged**
